{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9a1f9407",
   "metadata": {},
   "source": [
    "# Exercise 4: Batch homology grouping and downstream QC analyses\n",
    "\n",
    "Following on the same strategy we employed in exercises 1 and 2 of first learning how to employ `acute` on a single image and then scaling up to batch image data we will now take what we have learned in exercise 3 for homology grouping with the `starscape` and `constella` workflow and scale this method up for use on batch image data.  Following the generation of serial ID number homology groups and assigning them to our acute plms we will then assay the accuracy of these results through the use of a Quality Control (QC) test for `constella` and discuss how these outputs should be interpreted.\n",
    "\n",
    "To begin lets load the libraries and other input files we'll need to proceed..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1cc1b0bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib widget\n",
    "from plantcv import plantcv as pcv\n",
    "import os\n",
    "import numpy as np\n",
    "import cv2\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "212c6a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PlantCV adjustable global parameters\n",
    "pcv.params.debug = \"plot\"\n",
    "pcv.params.text_size = 20\n",
    "pcv.params.text_thickness = 10\n",
    "pcv.params.line_thickness = 14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f3428f8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Workflow parameters\n",
    "matplotlib.rcParams[\"figure.max_open_warning\"] = 50\n",
    "win = 25\n",
    "thresh = 90\n",
    "\n",
    "path = './imgs'\n",
    "days = range(10, 14)\n",
    "name_prefix = 'B100_rep1_d'\n",
    "\n",
    "outpath = './output'\n",
    "os.makedirs(outpath, exist_ok=True)\n",
    "outfile_prefix = os.path.join(outpath, 'B100_test')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7b0dd80",
   "metadata": {},
   "source": [
    "Now with our libraries loaded and initial parameters assigned let us begin by running a batch workflow with `acute` to generate our list of plms as a landmark dataframe.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a3715c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: roi_type='largest' will only return the largest contour and its immediate children. Other subcontours will be dropped.\n",
      "Contour volume: 2184.140382528305\n",
      "Fusing contour edges\n",
      "route C\n",
      "Landmark site: 1841, Start site: 1819, Term. site: 21\n",
      "Landmark point indices: [1841]\n",
      "Starting site indices: [1819]\n",
      "Termination site indices: [21]\n",
      "route C\n",
      "Landmark site: 308, Start site: 284, Term. site: 333\n",
      "Landmark point indices: [1841, 308]\n",
      "Starting site indices: [1819, 284]\n",
      "Termination site indices: [21, 333]\n",
      "route C\n",
      "Landmark site: 553, Start site: 528, Term. site: 577\n",
      "Landmark point indices: [1841, 308, 553]\n",
      "Starting site indices: [1819, 284, 528]\n",
      "Termination site indices: [21, 333, 577]\n",
      "route C\n",
      "Landmark site: 883, Start site: 869, Term. site: 894\n",
      "Landmark point indices: [1841, 308, 553, 883]\n",
      "Starting site indices: [1819, 284, 528, 869]\n",
      "Termination site indices: [21, 333, 577, 894]\n",
      "route C\n",
      "Landmark site: 976, Start site: 948, Term. site: 1001\n",
      "Landmark point indices: [1841, 308, 553, 883, 976]\n",
      "Starting site indices: [1819, 284, 528, 869, 948]\n",
      "Termination site indices: [21, 333, 577, 894, 1001]\n",
      "route C\n",
      "Landmark site: 1110, Start site: 1080, Term. site: 1137\n",
      "Landmark point indices: [1841, 308, 553, 883, 976, 1110]\n",
      "Starting site indices: [1819, 284, 528, 869, 948, 1080]\n",
      "Termination site indices: [21, 333, 577, 894, 1001, 1137]\n",
      "route C\n",
      "Landmark site: 1334, Start site: 1310, Term. site: 1359\n",
      "Landmark point indices: [1841, 308, 553, 883, 976, 1110, 1334]\n",
      "Starting site indices: [1819, 284, 528, 869, 948, 1080, 1310]\n",
      "Termination site indices: [21, 333, 577, 894, 1001, 1137, 1359]\n",
      "route C\n",
      "Landmark site: 1476, Start site: 1454, Term. site: 1502\n",
      "Landmark point indices: [1841, 308, 553, 883, 976, 1110, 1334, 1476]\n",
      "Starting site indices: [1819, 284, 528, 869, 948, 1080, 1310, 1454]\n",
      "Termination site indices: [21, 333, 577, 894, 1001, 1137, 1359, 1502]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fbf820aef27148a48545d9a15e17c57c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    landmark number: 8\n",
      "Warning: roi_type='largest' will only return the largest contour and its immediate children. Other subcontours will be dropped.\n",
      "Contour volume: 2436.750554203987\n",
      "Fusing contour edges\n",
      "route C\n",
      "Landmark site: 2070, Start site: 2047, Term. site: 18\n",
      "Landmark point indices: [2070]\n",
      "Starting site indices: [2047]\n",
      "Termination site indices: [18]\n",
      "route C\n",
      "Landmark site: 265, Start site: 241, Term. site: 289\n",
      "Landmark point indices: [2070, 265]\n",
      "Starting site indices: [2047, 241]\n",
      "Termination site indices: [18, 289]\n",
      "route C\n",
      "Landmark site: 401, Start site: 377, Term. site: 428\n",
      "Landmark point indices: [2070, 265, 401]\n",
      "Starting site indices: [2047, 241, 377]\n",
      "Termination site indices: [18, 289, 428]\n",
      "route C\n",
      "Landmark site: 598, Start site: 574, Term. site: 620\n",
      "Landmark point indices: [2070, 265, 401, 598]\n",
      "Starting site indices: [2047, 241, 377, 574]\n",
      "Termination site indices: [18, 289, 428, 620]\n",
      "route C\n",
      "Landmark site: 839, Start site: 816, Term. site: 862\n",
      "Landmark point indices: [2070, 265, 401, 598, 839]\n",
      "Starting site indices: [2047, 241, 377, 574, 816]\n",
      "Termination site indices: [18, 289, 428, 620, 862]\n",
      "route C\n",
      "Landmark site: 1159, Start site: 1143, Term. site: 1173\n",
      "Landmark point indices: [2070, 265, 401, 598, 839, 1159]\n",
      "Starting site indices: [2047, 241, 377, 574, 816, 1143]\n",
      "Termination site indices: [18, 289, 428, 620, 862, 1173]\n",
      "route C\n",
      "Landmark site: 1258, Start site: 1228, Term. site: 1282\n",
      "Landmark point indices: [2070, 265, 401, 598, 839, 1159, 1258]\n",
      "Starting site indices: [2047, 241, 377, 574, 816, 1143, 1228]\n",
      "Termination site indices: [18, 289, 428, 620, 862, 1173, 1282]\n",
      "route C\n",
      "Landmark site: 1383, Start site: 1352, Term. site: 1406\n",
      "Landmark point indices: [2070, 265, 401, 598, 839, 1159, 1258, 1383]\n",
      "Starting site indices: [2047, 241, 377, 574, 816, 1143, 1228, 1352]\n",
      "Termination site indices: [18, 289, 428, 620, 862, 1173, 1282, 1406]\n",
      "route C\n",
      "Landmark site: 1581, Start site: 1557, Term. site: 1605\n",
      "Landmark point indices: [2070, 265, 401, 598, 839, 1159, 1258, 1383, 1581]\n",
      "Starting site indices: [2047, 241, 377, 574, 816, 1143, 1228, 1352, 1557]\n",
      "Termination site indices: [18, 289, 428, 620, 862, 1173, 1282, 1406, 1605]\n",
      "route C\n",
      "Landmark site: 1707, Start site: 1692, Term. site: 1724\n",
      "Landmark point indices: [2070, 265, 401, 598, 839, 1159, 1258, 1383, 1581, 1707]\n",
      "Starting site indices: [2047, 241, 377, 574, 816, 1143, 1228, 1352, 1557, 1692]\n",
      "Termination site indices: [18, 289, 428, 620, 862, 1173, 1282, 1406, 1605, 1724]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c492abe1ea594a2797d9803852a7c715",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    landmark number: 10\n",
      "Warning: roi_type='largest' will only return the largest contour and its immediate children. Other subcontours will be dropped.\n",
      "Contour volume: 2664.7160643339157\n",
      "Fusing contour edges\n",
      "route C\n",
      "Landmark site: 0, Start site: 2242, Term. site: 21\n",
      "Landmark point indices: [0]\n",
      "Starting site indices: [2242]\n",
      "Termination site indices: [21]\n",
      "route C\n",
      "Landmark site: 280, Start site: 258, Term. site: 303\n",
      "Landmark point indices: [0, 280]\n",
      "Starting site indices: [2242, 258]\n",
      "Termination site indices: [21, 303]\n",
      "route C\n",
      "Landmark site: 494, Start site: 472, Term. site: 519\n",
      "Landmark point indices: [0, 280, 494]\n",
      "Starting site indices: [2242, 258, 472]\n",
      "Termination site indices: [21, 303, 519]\n",
      "route C\n",
      "Landmark site: 773, Start site: 750, Term. site: 799\n",
      "Landmark point indices: [0, 280, 494, 773]\n",
      "Starting site indices: [2242, 258, 472, 750]\n",
      "Termination site indices: [21, 303, 519, 799]\n",
      "route C\n",
      "Landmark site: 1008, Start site: 988, Term. site: 1031\n",
      "Landmark point indices: [0, 280, 494, 773, 1008]\n",
      "Starting site indices: [2242, 258, 472, 750, 988]\n",
      "Termination site indices: [21, 303, 519, 799, 1031]\n",
      "route C\n",
      "Landmark site: 1323, Start site: 1311, Term. site: 1338\n",
      "Landmark point indices: [0, 280, 494, 773, 1008, 1323]\n",
      "Starting site indices: [2242, 258, 472, 750, 988, 1311]\n",
      "Termination site indices: [21, 303, 519, 799, 1031, 1338]\n",
      "route C\n",
      "Landmark site: 1408, Start site: 1383, Term. site: 1431\n",
      "Landmark point indices: [0, 280, 494, 773, 1008, 1323, 1408]\n",
      "Starting site indices: [2242, 258, 472, 750, 988, 1311, 1383]\n",
      "Termination site indices: [21, 303, 519, 799, 1031, 1338, 1431]\n",
      "route C\n",
      "Landmark site: 1517, Start site: 1487, Term. site: 1544\n",
      "Landmark point indices: [0, 280, 494, 773, 1008, 1323, 1408, 1517]\n",
      "Starting site indices: [2242, 258, 472, 750, 988, 1311, 1383, 1487]\n",
      "Termination site indices: [21, 303, 519, 799, 1031, 1338, 1431, 1544]\n",
      "route C\n",
      "Landmark site: 1733, Start site: 1702, Term. site: 1757\n",
      "Landmark point indices: [0, 280, 494, 773, 1008, 1323, 1408, 1517, 1733]\n",
      "Starting site indices: [2242, 258, 472, 750, 988, 1311, 1383, 1487, 1702]\n",
      "Termination site indices: [21, 303, 519, 799, 1031, 1338, 1431, 1544, 1757]\n",
      "route C\n",
      "Landmark site: 1873, Start site: 1856, Term. site: 1888\n",
      "Landmark point indices: [0, 280, 494, 773, 1008, 1323, 1408, 1517, 1733, 1873]\n",
      "Starting site indices: [2242, 258, 472, 750, 988, 1311, 1383, 1487, 1702, 1856]\n",
      "Termination site indices: [21, 303, 519, 799, 1031, 1338, 1431, 1544, 1757, 1888]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "15baae725aeb4e88bdacd71565ab37d2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    landmark number: 10\n",
      "Warning: roi_type='largest' will only return the largest contour and its immediate children. Other subcontours will be dropped.\n",
      "Contour volume: 2932.647084593773\n",
      "Fusing contour edges\n",
      "route C\n",
      "Landmark site: 0, Start site: 2442, Term. site: 22\n",
      "Landmark point indices: [0]\n",
      "Starting site indices: [2442]\n",
      "Termination site indices: [22]\n",
      "route C\n",
      "Landmark site: 358, Start site: 333, Term. site: 386\n",
      "Landmark point indices: [0, 358]\n",
      "Starting site indices: [2442, 333]\n",
      "Termination site indices: [22, 386]\n",
      "route C\n",
      "Landmark site: 595, Start site: 573, Term. site: 617\n",
      "Landmark point indices: [0, 358, 595]\n",
      "Starting site indices: [2442, 333, 573]\n",
      "Termination site indices: [22, 386, 617]\n",
      "route C\n",
      "Landmark site: 926, Start site: 905, Term. site: 943\n",
      "Landmark point indices: [0, 358, 595, 926]\n",
      "Starting site indices: [2442, 333, 573, 905]\n",
      "Termination site indices: [22, 386, 617, 943]\n",
      "route C\n",
      "Landmark site: 1004, Start site: 978, Term. site: 1027\n",
      "Landmark point indices: [0, 358, 595, 926, 1004]\n",
      "Starting site indices: [2442, 333, 573, 905, 978]\n",
      "Termination site indices: [22, 386, 617, 943, 1027]\n",
      "route C\n",
      "Landmark site: 1110, Start site: 1079, Term. site: 1133\n",
      "Landmark point indices: [0, 358, 595, 926, 1004, 1110]\n",
      "Starting site indices: [2442, 333, 573, 905, 978, 1079]\n",
      "Termination site indices: [22, 386, 617, 943, 1027, 1133]\n",
      "route C\n",
      "Landmark site: 1312, Start site: 1287, Term. site: 1339\n",
      "Landmark point indices: [0, 358, 595, 926, 1004, 1110, 1312]\n",
      "Starting site indices: [2442, 333, 573, 905, 978, 1079, 1287]\n",
      "Termination site indices: [22, 386, 617, 943, 1027, 1133, 1339]\n",
      "route C\n",
      "Landmark site: 1441, Start site: 1426, Term. site: 1458\n",
      "Landmark point indices: [0, 358, 595, 926, 1004, 1110, 1312, 1441]\n",
      "Starting site indices: [2442, 333, 573, 905, 978, 1079, 1287, 1426]\n",
      "Termination site indices: [22, 386, 617, 943, 1027, 1133, 1339, 1458]\n",
      "route C\n",
      "Landmark site: 1857, Start site: 1837, Term. site: 1880\n",
      "Landmark point indices: [0, 358, 595, 926, 1004, 1110, 1312, 1441, 1857]\n",
      "Starting site indices: [2442, 333, 573, 905, 978, 1079, 1287, 1426, 1837]\n",
      "Termination site indices: [22, 386, 617, 943, 1027, 1133, 1339, 1458, 1880]\n",
      "route C\n",
      "Landmark site: 2163, Start site: 2137, Term. site: 2187\n",
      "Landmark point indices: [0, 358, 595, 926, 1004, 1110, 1312, 1441, 1857, 2163]\n",
      "Starting site indices: [2442, 333, 573, 905, 978, 1079, 1287, 1426, 1837, 2137]\n",
      "Termination site indices: [22, 386, 617, 943, 1027, 1133, 1339, 1458, 1880, 2187]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9d13322508f2408b8dd15866ed601bf8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    landmark number: 10\n"
     ]
    }
   ],
   "source": [
    "landmark_output=[['group', 'plmname', 'filename', 'plm_x', 'plm_y', 'SS_x', 'SS_y', 'TS_x', 'TS_y', 'CC_ratio']]\n",
    "for day in days:\n",
    "    # Turn off debugging outputs for most steps to avoid a huge amount of output\n",
    "    pcv.params.debug = None\n",
    "    # 1. Reading our image into the environment\n",
    "    img, imgpath, imgname = pcv.readimage(filename=os.path.join(path, f\"{name_prefix}{day}.jpg\"))\n",
    "    \n",
    "    # 2. Converting our RGB image into an Lab color space and extract the a and b channels\n",
    "    img_a = pcv.rgb2gray_lab(rgb_img=img, channel=\"a\")\n",
    "    img_b = pcv.rgb2gray_lab(rgb_img=img, channel=\"b\")\n",
    "    \n",
    "    # 3. Thresholding our a and b color channels to create two masks\n",
    "    mask_a = pcv.threshold.binary(gray_img=img_a, threshold=123, max_value=255, object_type=\"dark\")\n",
    "    mask_b = pcv.threshold.binary(gray_img=img_b, threshold=133, max_value=255, object_type=\"light\")\n",
    "    \n",
    "    # 4. Merging our individual a and b thresholded masks\n",
    "    mask = pcv.logical_and(bin_img1=mask_a, bin_img2=mask_b)\n",
    "    \n",
    "    # 5. Extracting our contours from the final mask\n",
    "    cnt, cnt_str = pcv.find_objects(img=img, mask=mask)\n",
    "    \n",
    "    # 6. Find largest contour of subject (outer boundary of subject)\n",
    "    roi, roi_str = pcv.roi.rectangle(img=img, x=500, y=500, h=1100, w=800)\n",
    "    kept_cnt, kept_str, kept_mask, kept_area = pcv.roi_objects(img=img, roi_contour=roi, roi_hierarchy=roi_str, \n",
    "                                                               object_contour=cnt, obj_hierarchy=cnt_str, roi_type=\"largest\")\n",
    "    \n",
    "    # 7. Extracting pseudo-landmarks from the plant contours\n",
    "    plt_img = np.copy(img)\n",
    "    for cont in kept_cnt:\n",
    "        if cv2.arcLength(cont, True) > 2*win:\n",
    "            print(f'Contour volume: {cv2.arcLength(cont, True)}')\n",
    "            # Turn debugging back on for acute\n",
    "            pcv.params.debug = \"plot\"\n",
    "            homolog_pts, homolog_start, homolog_stop, homolog_cc, chain, verbose = pcv.homology.acute(img=img, obj=cont, mask=mask,\n",
    "                                                                                                      win=win, threshold=thresh)\n",
    "            print(f'    landmark number: {len(homolog_pts)}')\n",
    "\n",
    "            for h in range(0, len(homolog_pts)):\n",
    "                landmark_output.append([None, f\"{name_prefix}{day}_plm{h+1}\", f\"{name_prefix}{day}\", homolog_pts[h][0][0], homolog_pts[h][0][1], \n",
    "                                        homolog_start[h][0][0], homolog_start[h][0][1], homolog_stop[h][0][0], homolog_stop[h][0][1], homolog_cc[h],])\n",
    "\n",
    "# Convert out output to a pandas dataframe for ease of use hereafter...\n",
    "landmark_pandas = pd.DataFrame(landmark_output[1:len(landmark_output)], columns=landmark_output[0][0:11])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d50d2bb6",
   "metadata": {},
   "source": [
    "Before we continue lets check our landmarks once more to ensure we'll have what we need to run our homology grouping workflow..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e263384",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>plmname</th>\n",
       "      <th>filename</th>\n",
       "      <th>plm_x</th>\n",
       "      <th>plm_y</th>\n",
       "      <th>SS_x</th>\n",
       "      <th>SS_y</th>\n",
       "      <th>TS_x</th>\n",
       "      <th>TS_y</th>\n",
       "      <th>CC_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>B100_rep1_d10_plm1</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>901</td>\n",
       "      <td>1151</td>\n",
       "      <td>892</td>\n",
       "      <td>1173</td>\n",
       "      <td>885</td>\n",
       "      <td>1167</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>B100_rep1_d10_plm2</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>787</td>\n",
       "      <td>1425</td>\n",
       "      <td>789</td>\n",
       "      <td>1401</td>\n",
       "      <td>773</td>\n",
       "      <td>1405</td>\n",
       "      <td>6.219512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>B100_rep1_d10_plm3</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>571</td>\n",
       "      <td>1344</td>\n",
       "      <td>594</td>\n",
       "      <td>1338</td>\n",
       "      <td>595</td>\n",
       "      <td>1342</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>B100_rep1_d10_plm4</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>793</td>\n",
       "      <td>1523</td>\n",
       "      <td>796</td>\n",
       "      <td>1511</td>\n",
       "      <td>783</td>\n",
       "      <td>1519</td>\n",
       "      <td>18.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>B100_rep1_d10_plm5</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>712</td>\n",
       "      <td>1511</td>\n",
       "      <td>736</td>\n",
       "      <td>1508</td>\n",
       "      <td>736</td>\n",
       "      <td>1512</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  group             plmname       filename  plm_x  plm_y  SS_x  SS_y  TS_x  \\\n",
       "0  None  B100_rep1_d10_plm1  B100_rep1_d10    901   1151   892  1173   885   \n",
       "1  None  B100_rep1_d10_plm2  B100_rep1_d10    787   1425   789  1401   773   \n",
       "2  None  B100_rep1_d10_plm3  B100_rep1_d10    571   1344   594  1338   595   \n",
       "3  None  B100_rep1_d10_plm4  B100_rep1_d10    793   1523   796  1511   783   \n",
       "4  None  B100_rep1_d10_plm5  B100_rep1_d10    712   1511   736  1508   736   \n",
       "\n",
       "   TS_y    CC_ratio  \n",
       "0  1167  255.000000  \n",
       "1  1405    6.219512  \n",
       "2  1342  255.000000  \n",
       "3  1519   18.888889  \n",
       "4  1512  255.000000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "landmark_pandas.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c679851",
   "metadata": {},
   "source": [
    "Provided the head of our table loaded in properly we're now ready to begin a batch run of our homology pipeline which follows the same structure as our previous exercise utilizing `starscape` and `constella`. Note that we have left out `space` from this workflow, studies of this pipelines accuracy have suggested that this function isn't essential for generating tangible improvements in homology grouping.  As such this workflow of parsing segmented morphological data directly into `starscape` is considered the present best practice of this approach. Notice how our groups within `landmark_pandas` are universally assigned to 'None'? On the other side of this code block we should see the results of transfering group IDs from paired frames to this original dataframe resulting in forthcoming changes to this columns values.  Let's get started! \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9dfdc80a-afe8-46d3-8474-7b857ff7ce96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "######################################################\n",
      "Beginning next iteration for days 10 and 11\n",
      "Running StarScape...\n",
      "Eigenvalues:  [3.74069468 2.73593046 0.92648202 0.00523252] \n",
      "\n",
      "\n",
      "Var. Explained:  [0.5046969  0.36913348 0.12500154 0.00070597] \n",
      "\n",
      "\n",
      "Cumul. Var. Explained:  [0.5046969  0.87383038 0.99883192 0.99953789] \n",
      "\n",
      "\n",
      "2  components sufficiently informative\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d8720cb2aab94ff888880f87de253dd5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "de240d2a7e484bb0983d01be7d47f479",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Constella...\n",
      "18 plms to group\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "773fedb4164948d9a2b364ae0031a421",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b0e1b1be90494ccc8adc2449d82dcb29",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "37a94c387d2b4a8abc0f3a78758d1fce",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "55e96f72e85f48b98cecbf3eeb9e5f96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "######################################################\n",
      "Beginning next iteration for days 11 and 12\n",
      "Running StarScape...\n",
      "Eigenvalues:  [3.70796764 2.74004966 0.91203917 0.00514027] \n",
      "\n",
      "\n",
      "Var. Explained:  [0.50322418 0.37186388 0.12377674 0.00069761] \n",
      "\n",
      "\n",
      "Cumul. Var. Explained:  [0.50322418 0.87508806 0.99886481 0.99956241] \n",
      "\n",
      "\n",
      "2  components sufficiently informative\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4fb5359b1ba34b7694cd5e61075aed2e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "40860f0697274088b480376657770c67",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Constella...\n",
      "20 plms to group\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "70a217964d804931984e805f5f64daca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "00e9c04bf81541599624735bcd9f012b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9897877b537d4b8c89a888fafe9b6cf2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7d0a69bee9de426986b26b859aabafa9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "######################################################\n",
      "Beginning next iteration for days 12 and 13\n",
      "Running StarScape...\n",
      "Eigenvalues:  [3.5310468  2.93567762 0.89436779 0.00450095] \n",
      "\n",
      "\n",
      "Var. Explained:  [0.47921349 0.39841339 0.12137849 0.00061084] \n",
      "\n",
      "\n",
      "Cumul. Var. Explained:  [0.47921349 0.87762689 0.99900537 0.99961622] \n",
      "\n",
      "\n",
      "2  components sufficiently informative\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "23f007d319df433c9c51db79809eba66",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "64e0cd33cf214b81a5dda448181fc3dc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Constella...\n",
      "20 plms to group\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c33cf6c35486436ea17ee1c196ebdaab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a531fbcb13f474db9e61eb956596900",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aae64f49bc9a43d39e57221316ca9d14",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fb61b2eccf25475d999a9e3be7f75f67",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "group_iter = 1\n",
    "# Get all the filenames from the dataframe for filtering\n",
    "filenames = landmark_pandas.loc[:,['filename']].values\n",
    "# Loop over the days we defined above\n",
    "for day in days[:-1]:\n",
    "    # Iteration is between day and day + 1\n",
    "    print(\"######################################################\")\n",
    "    print(f'Beginning next iteration for days {day} and {day + 1}')\n",
    "    \n",
    "    # Current data frame of pseudo-landmarks from the pair of day and day + 1\n",
    "    groupA = f\"{name_prefix}{day}\"\n",
    "    groupB = f\"{name_prefix}{day + 1}\"\n",
    "    cur_plms = landmark_pandas[filenames == groupA]\n",
    "    cur_plms = cur_plms.append(landmark_pandas[filenames == groupB])\n",
    "    cur_plms = cur_plms.reset_index(drop=True)\n",
    "    \n",
    "    print('Running StarScape...')\n",
    "    finalDf, eigenvals, loadings = pcv.homology.starscape(cur_plms=cur_plms, group_a=groupA, group_b=groupB, outfile_prefix=f\"{outfile_prefix}_{day}\")\n",
    "    \n",
    "    print('Running Constella...')\n",
    "    cur_plms, group_iter = pcv.homology.constella(cur_plms=cur_plms, pc_starscape=finalDf, group_iter=group_iter, outfile_prefix=f\"{outfile_prefix}_{day}\")\n",
    "    \n",
    "    plmnames = landmark_pandas.loc[:,['plmname']].values\n",
    "    cur_plmnames = cur_plms.loc[:,['plmname']].values\n",
    "    \n",
    "    for name in cur_plmnames:\n",
    "        landmark_index = [i for i, x in enumerate(plmnames == name) if x]\n",
    "        cur_plms_index = [i for i, x in enumerate(cur_plmnames == name) if x]\n",
    "        if landmark_pandas.iloc[landmark_index, 0].values == None:\n",
    "            landmark_pandas.iloc[landmark_index, 0] = cur_plms.iloc[cur_plms_index, 0]\n",
    "    \n",
    "    img1, _, _ = pcv.readimage(filename=os.path.join(path, f\"{name_prefix}{day}.jpg\"))\n",
    "    \n",
    "    for p in range(0, cur_plms.shape[0]):\n",
    "        if f\"{name_prefix}{day}\" in cur_plms.at[p, 'plmname']:        \n",
    "            cv2.putText(img1, str(cur_plms.at[p, 'group']), \n",
    "                        (int(cur_plms.at[p, 'plm_x'])-10, int(cur_plms.at[p, 'plm_y'])), \n",
    "                        cv2.FONT_ITALIC, 1.5, (255,0,0), 6)\n",
    "    \n",
    "    img2, _, _ = pcv.readimage(filename=os.path.join(path, f\"{name_prefix}{day + 1}.jpg\"))\n",
    "\n",
    "    for p in range(0, cur_plms.shape[0]):\n",
    "        if f\"{name_prefix}{day + 1}\" in cur_plms.at[p, 'plmname']:        \n",
    "            cv2.putText(img2, str(cur_plms.at[p, 'group']), \n",
    "                        (int(cur_plms.at[p, 'plm_x'])-10, int(cur_plms.at[p, 'plm_y'])), \n",
    "                        cv2.FONT_ITALIC, 1.5, (0,0,255), 6)\n",
    "    img_neighbors = cv2.hconcat((img1, img2))\n",
    "\n",
    "    plm_groups_fig=plt.figure(figsize=(8, 6))\n",
    "    plm_groups_fig=plt.imshow(img_neighbors)\n",
    "    plm_groups_fig=plt.xscale('linear')\n",
    "    plm_groups_fig=plt.axis('off')\n",
    "    plm_groups_fig=plt.title(f'B100 day {day}-{day + 1}')\n",
    "    plt.show(plm_groups_fig)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4fdc6e-ec60-41fb-a485-c63097592c53",
   "metadata": {},
   "source": [
    "With the homology grouping workflow now completed a decent array of graphical outputs should be visible above displaying not only our PCA related graphs and the dendrogram used for our hierarchical clustering at each step, but also side by side images of our plants as well as their labeled homology groups to enable for easy point of reference for calling accuracy.  Let's have a look at our de novo homology groups on our original `landmark_pandas` dataframe, this time we'll have a look at the full table though instead of just taking a quick glance at the head.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "75a3567a-bdb8-4ffc-8857-88c1071165c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>plmname</th>\n",
       "      <th>filename</th>\n",
       "      <th>plm_x</th>\n",
       "      <th>plm_y</th>\n",
       "      <th>SS_x</th>\n",
       "      <th>SS_y</th>\n",
       "      <th>TS_x</th>\n",
       "      <th>TS_y</th>\n",
       "      <th>CC_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>B100_rep1_d10_plm1</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>901</td>\n",
       "      <td>1151</td>\n",
       "      <td>892</td>\n",
       "      <td>1173</td>\n",
       "      <td>885</td>\n",
       "      <td>1167</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>B100_rep1_d10_plm2</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>787</td>\n",
       "      <td>1425</td>\n",
       "      <td>789</td>\n",
       "      <td>1401</td>\n",
       "      <td>773</td>\n",
       "      <td>1405</td>\n",
       "      <td>6.219512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>B100_rep1_d10_plm3</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>571</td>\n",
       "      <td>1344</td>\n",
       "      <td>594</td>\n",
       "      <td>1338</td>\n",
       "      <td>595</td>\n",
       "      <td>1342</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>B100_rep1_d10_plm4</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>793</td>\n",
       "      <td>1523</td>\n",
       "      <td>796</td>\n",
       "      <td>1511</td>\n",
       "      <td>783</td>\n",
       "      <td>1519</td>\n",
       "      <td>18.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>B100_rep1_d10_plm5</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>712</td>\n",
       "      <td>1511</td>\n",
       "      <td>736</td>\n",
       "      <td>1508</td>\n",
       "      <td>736</td>\n",
       "      <td>1512</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>B100_rep1_d10_plm6</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>803</td>\n",
       "      <td>1555</td>\n",
       "      <td>795</td>\n",
       "      <td>1533</td>\n",
       "      <td>807</td>\n",
       "      <td>1532</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>B100_rep1_d10_plm7</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>922</td>\n",
       "      <td>1415</td>\n",
       "      <td>898</td>\n",
       "      <td>1420</td>\n",
       "      <td>898</td>\n",
       "      <td>1416</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>B100_rep1_d10_plm8</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>803</td>\n",
       "      <td>1478</td>\n",
       "      <td>816</td>\n",
       "      <td>1459</td>\n",
       "      <td>801</td>\n",
       "      <td>1454</td>\n",
       "      <td>12.207447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>B100_rep1_d11_plm1</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>912</td>\n",
       "      <td>1123</td>\n",
       "      <td>904</td>\n",
       "      <td>1146</td>\n",
       "      <td>895</td>\n",
       "      <td>1139</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>B100_rep1_d11_plm2</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>786</td>\n",
       "      <td>1370</td>\n",
       "      <td>794</td>\n",
       "      <td>1347</td>\n",
       "      <td>784</td>\n",
       "      <td>1346</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>B100_rep1_d11_plm3</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>752</td>\n",
       "      <td>1236</td>\n",
       "      <td>761</td>\n",
       "      <td>1258</td>\n",
       "      <td>754</td>\n",
       "      <td>1260</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7</td>\n",
       "      <td>B100_rep1_d11_plm4</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>783</td>\n",
       "      <td>1427</td>\n",
       "      <td>781</td>\n",
       "      <td>1403</td>\n",
       "      <td>770</td>\n",
       "      <td>1406</td>\n",
       "      <td>32.692308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5</td>\n",
       "      <td>B100_rep1_d11_plm5</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>576</td>\n",
       "      <td>1353</td>\n",
       "      <td>598</td>\n",
       "      <td>1342</td>\n",
       "      <td>599</td>\n",
       "      <td>1347</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>B100_rep1_d11_plm6</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>790</td>\n",
       "      <td>1525</td>\n",
       "      <td>791</td>\n",
       "      <td>1509</td>\n",
       "      <td>780</td>\n",
       "      <td>1517</td>\n",
       "      <td>23.448276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4</td>\n",
       "      <td>B100_rep1_d11_plm7</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>706</td>\n",
       "      <td>1510</td>\n",
       "      <td>729</td>\n",
       "      <td>1505</td>\n",
       "      <td>730</td>\n",
       "      <td>1509</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>B100_rep1_d11_plm8</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>802</td>\n",
       "      <td>1553</td>\n",
       "      <td>791</td>\n",
       "      <td>1532</td>\n",
       "      <td>801</td>\n",
       "      <td>1530</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>6</td>\n",
       "      <td>B100_rep1_d11_plm9</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>914</td>\n",
       "      <td>1417</td>\n",
       "      <td>891</td>\n",
       "      <td>1423</td>\n",
       "      <td>890</td>\n",
       "      <td>1417</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>3</td>\n",
       "      <td>B100_rep1_d11_plm10</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>798</td>\n",
       "      <td>1475</td>\n",
       "      <td>809</td>\n",
       "      <td>1464</td>\n",
       "      <td>795</td>\n",
       "      <td>1458</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>8</td>\n",
       "      <td>B100_rep1_d12_plm1</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>932</td>\n",
       "      <td>1108</td>\n",
       "      <td>925</td>\n",
       "      <td>1131</td>\n",
       "      <td>916</td>\n",
       "      <td>1124</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>9</td>\n",
       "      <td>B100_rep1_d12_plm2</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>784</td>\n",
       "      <td>1362</td>\n",
       "      <td>793</td>\n",
       "      <td>1340</td>\n",
       "      <td>780</td>\n",
       "      <td>1339</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>10</td>\n",
       "      <td>B100_rep1_d12_plm3</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>720</td>\n",
       "      <td>1151</td>\n",
       "      <td>732</td>\n",
       "      <td>1172</td>\n",
       "      <td>724</td>\n",
       "      <td>1175</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>7</td>\n",
       "      <td>B100_rep1_d12_plm4</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>779</td>\n",
       "      <td>1425</td>\n",
       "      <td>779</td>\n",
       "      <td>1402</td>\n",
       "      <td>763</td>\n",
       "      <td>1407</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5</td>\n",
       "      <td>B100_rep1_d12_plm5</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>568</td>\n",
       "      <td>1367</td>\n",
       "      <td>588</td>\n",
       "      <td>1356</td>\n",
       "      <td>591</td>\n",
       "      <td>1361</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2</td>\n",
       "      <td>B100_rep1_d12_plm6</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>790</td>\n",
       "      <td>1523</td>\n",
       "      <td>789</td>\n",
       "      <td>1511</td>\n",
       "      <td>778</td>\n",
       "      <td>1518</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>4</td>\n",
       "      <td>B100_rep1_d12_plm7</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>713</td>\n",
       "      <td>1509</td>\n",
       "      <td>736</td>\n",
       "      <td>1506</td>\n",
       "      <td>736</td>\n",
       "      <td>1511</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>B100_rep1_d12_plm8</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>801</td>\n",
       "      <td>1555</td>\n",
       "      <td>790</td>\n",
       "      <td>1534</td>\n",
       "      <td>801</td>\n",
       "      <td>1533</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>6</td>\n",
       "      <td>B100_rep1_d12_plm9</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>919</td>\n",
       "      <td>1422</td>\n",
       "      <td>896</td>\n",
       "      <td>1425</td>\n",
       "      <td>895</td>\n",
       "      <td>1422</td>\n",
       "      <td>249.900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3</td>\n",
       "      <td>B100_rep1_d12_plm10</td>\n",
       "      <td>B100_rep1_d12</td>\n",
       "      <td>797</td>\n",
       "      <td>1475</td>\n",
       "      <td>809</td>\n",
       "      <td>1465</td>\n",
       "      <td>793</td>\n",
       "      <td>1460</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>10</td>\n",
       "      <td>B100_rep1_d13_plm1</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>659</td>\n",
       "      <td>1079</td>\n",
       "      <td>674</td>\n",
       "      <td>1098</td>\n",
       "      <td>670</td>\n",
       "      <td>1101</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>7</td>\n",
       "      <td>B100_rep1_d13_plm2</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>771</td>\n",
       "      <td>1428</td>\n",
       "      <td>770</td>\n",
       "      <td>1404</td>\n",
       "      <td>755</td>\n",
       "      <td>1410</td>\n",
       "      <td>24.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>5</td>\n",
       "      <td>B100_rep1_d13_plm3</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>560</td>\n",
       "      <td>1373</td>\n",
       "      <td>581</td>\n",
       "      <td>1362</td>\n",
       "      <td>582</td>\n",
       "      <td>1367</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>2</td>\n",
       "      <td>B100_rep1_d13_plm4</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>784</td>\n",
       "      <td>1526</td>\n",
       "      <td>784</td>\n",
       "      <td>1506</td>\n",
       "      <td>771</td>\n",
       "      <td>1518</td>\n",
       "      <td>2.198276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>4</td>\n",
       "      <td>B100_rep1_d13_plm5</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>714</td>\n",
       "      <td>1512</td>\n",
       "      <td>737</td>\n",
       "      <td>1506</td>\n",
       "      <td>737</td>\n",
       "      <td>1513</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>B100_rep1_d13_plm6</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>798</td>\n",
       "      <td>1557</td>\n",
       "      <td>784</td>\n",
       "      <td>1538</td>\n",
       "      <td>796</td>\n",
       "      <td>1535</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>6</td>\n",
       "      <td>B100_rep1_d13_plm7</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>904</td>\n",
       "      <td>1420</td>\n",
       "      <td>882</td>\n",
       "      <td>1428</td>\n",
       "      <td>881</td>\n",
       "      <td>1424</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>3</td>\n",
       "      <td>B100_rep1_d13_plm8</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>789</td>\n",
       "      <td>1480</td>\n",
       "      <td>801</td>\n",
       "      <td>1469</td>\n",
       "      <td>786</td>\n",
       "      <td>1463</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>8</td>\n",
       "      <td>B100_rep1_d13_plm9</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>971</td>\n",
       "      <td>1115</td>\n",
       "      <td>957</td>\n",
       "      <td>1133</td>\n",
       "      <td>950</td>\n",
       "      <td>1125</td>\n",
       "      <td>255.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>9</td>\n",
       "      <td>B100_rep1_d13_plm10</td>\n",
       "      <td>B100_rep1_d13</td>\n",
       "      <td>774</td>\n",
       "      <td>1365</td>\n",
       "      <td>781</td>\n",
       "      <td>1342</td>\n",
       "      <td>772</td>\n",
       "      <td>1341</td>\n",
       "      <td>64.830508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   group              plmname       filename  plm_x  plm_y  SS_x  SS_y  TS_x  \\\n",
       "0      8   B100_rep1_d10_plm1  B100_rep1_d10    901   1151   892  1173   885   \n",
       "1      7   B100_rep1_d10_plm2  B100_rep1_d10    787   1425   789  1401   773   \n",
       "2      5   B100_rep1_d10_plm3  B100_rep1_d10    571   1344   594  1338   595   \n",
       "3      2   B100_rep1_d10_plm4  B100_rep1_d10    793   1523   796  1511   783   \n",
       "4      4   B100_rep1_d10_plm5  B100_rep1_d10    712   1511   736  1508   736   \n",
       "5      1   B100_rep1_d10_plm6  B100_rep1_d10    803   1555   795  1533   807   \n",
       "6      6   B100_rep1_d10_plm7  B100_rep1_d10    922   1415   898  1420   898   \n",
       "7      3   B100_rep1_d10_plm8  B100_rep1_d10    803   1478   816  1459   801   \n",
       "8      8   B100_rep1_d11_plm1  B100_rep1_d11    912   1123   904  1146   895   \n",
       "9      9   B100_rep1_d11_plm2  B100_rep1_d11    786   1370   794  1347   784   \n",
       "10    10   B100_rep1_d11_plm3  B100_rep1_d11    752   1236   761  1258   754   \n",
       "11     7   B100_rep1_d11_plm4  B100_rep1_d11    783   1427   781  1403   770   \n",
       "12     5   B100_rep1_d11_plm5  B100_rep1_d11    576   1353   598  1342   599   \n",
       "13     2   B100_rep1_d11_plm6  B100_rep1_d11    790   1525   791  1509   780   \n",
       "14     4   B100_rep1_d11_plm7  B100_rep1_d11    706   1510   729  1505   730   \n",
       "15     1   B100_rep1_d11_plm8  B100_rep1_d11    802   1553   791  1532   801   \n",
       "16     6   B100_rep1_d11_plm9  B100_rep1_d11    914   1417   891  1423   890   \n",
       "17     3  B100_rep1_d11_plm10  B100_rep1_d11    798   1475   809  1464   795   \n",
       "18     8   B100_rep1_d12_plm1  B100_rep1_d12    932   1108   925  1131   916   \n",
       "19     9   B100_rep1_d12_plm2  B100_rep1_d12    784   1362   793  1340   780   \n",
       "20    10   B100_rep1_d12_plm3  B100_rep1_d12    720   1151   732  1172   724   \n",
       "21     7   B100_rep1_d12_plm4  B100_rep1_d12    779   1425   779  1402   763   \n",
       "22     5   B100_rep1_d12_plm5  B100_rep1_d12    568   1367   588  1356   591   \n",
       "23     2   B100_rep1_d12_plm6  B100_rep1_d12    790   1523   789  1511   778   \n",
       "24     4   B100_rep1_d12_plm7  B100_rep1_d12    713   1509   736  1506   736   \n",
       "25     1   B100_rep1_d12_plm8  B100_rep1_d12    801   1555   790  1534   801   \n",
       "26     6   B100_rep1_d12_plm9  B100_rep1_d12    919   1422   896  1425   895   \n",
       "27     3  B100_rep1_d12_plm10  B100_rep1_d12    797   1475   809  1465   793   \n",
       "28    10   B100_rep1_d13_plm1  B100_rep1_d13    659   1079   674  1098   670   \n",
       "29     7   B100_rep1_d13_plm2  B100_rep1_d13    771   1428   770  1404   755   \n",
       "30     5   B100_rep1_d13_plm3  B100_rep1_d13    560   1373   581  1362   582   \n",
       "31     2   B100_rep1_d13_plm4  B100_rep1_d13    784   1526   784  1506   771   \n",
       "32     4   B100_rep1_d13_plm5  B100_rep1_d13    714   1512   737  1506   737   \n",
       "33     1   B100_rep1_d13_plm6  B100_rep1_d13    798   1557   784  1538   796   \n",
       "34     6   B100_rep1_d13_plm7  B100_rep1_d13    904   1420   882  1428   881   \n",
       "35     3   B100_rep1_d13_plm8  B100_rep1_d13    789   1480   801  1469   786   \n",
       "36     8   B100_rep1_d13_plm9  B100_rep1_d13    971   1115   957  1133   950   \n",
       "37     9  B100_rep1_d13_plm10  B100_rep1_d13    774   1365   781  1342   772   \n",
       "\n",
       "    TS_y    CC_ratio  \n",
       "0   1167  255.000000  \n",
       "1   1405    6.219512  \n",
       "2   1342  255.000000  \n",
       "3   1519   18.888889  \n",
       "4   1512  255.000000  \n",
       "5   1532  255.000000  \n",
       "6   1416  255.000000  \n",
       "7   1454   12.207447  \n",
       "8   1139  255.000000  \n",
       "9   1346    0.000000  \n",
       "10  1260  255.000000  \n",
       "11  1406   32.692308  \n",
       "12  1347  255.000000  \n",
       "13  1517   23.448276  \n",
       "14  1509  255.000000  \n",
       "15  1530  255.000000  \n",
       "16  1417  255.000000  \n",
       "17  1458    0.000000  \n",
       "18  1124  255.000000  \n",
       "19  1339    0.000000  \n",
       "20  1175  255.000000  \n",
       "21  1407    0.000000  \n",
       "22  1361  255.000000  \n",
       "23  1518    0.000000  \n",
       "24  1511  255.000000  \n",
       "25  1533  255.000000  \n",
       "26  1422  249.900000  \n",
       "27  1460    0.000000  \n",
       "28  1101  255.000000  \n",
       "29  1410   24.000000  \n",
       "30  1367  255.000000  \n",
       "31  1518    2.198276  \n",
       "32  1513  255.000000  \n",
       "33  1535  255.000000  \n",
       "34  1424  255.000000  \n",
       "35  1463    0.000000  \n",
       "36  1125  255.000000  \n",
       "37  1341   64.830508  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "landmark_pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ce0cf09-050e-41ba-bc33-b8570b5e7aa4",
   "metadata": {},
   "source": [
    "We see group serial numbers 1-10 repeating once for each frame so it appears things ran pretty well! Moreover, when we glance at the side-by-side images with the serial numbers superimposed onto the original images things look like they're grouping as we'd expect.  \n",
    "\n",
    "However, as with all de novo methods there is the possibility for errors to be introduced which we might miss at a glance.  This brings us to a key aspect of our plm workflow when scaling up to a full sized project which is Quality Control (QC) assessment of our de novo homologies.  This is often done with a reduced subset of our full dataset in order to give us a general idea of the overall accuracy of our calls.  Although there is currently one method of producing input `starscape` files to feeding to `constella` eventually as other ways to rescale our `acute` outputs are developed this method can provide a helpful means of comparing what method of metadata generation (plmSpace) and multivariate space transformation (`starscape`) is the best for maximizing biologically informative signal.  With this being said let's begin by loading in a table of our landmarks which have been annotated to represent the biological structures they represent*.\n",
    "\n",
    "*Although not seen in this situation it is common practice to denote random plms which don't represent any meaningful features as '-'.*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bc669113-c931-4df4-9ac8-2466d7a3eef6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>plmname</th>\n",
       "      <th>filename</th>\n",
       "      <th>plm_x</th>\n",
       "      <th>plm_y</th>\n",
       "      <th>SS_x</th>\n",
       "      <th>SS_y</th>\n",
       "      <th>TS_x</th>\n",
       "      <th>TS_y</th>\n",
       "      <th>CC_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>leaf5</td>\n",
       "      <td>B100_rep1_d10_plm1</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>901</td>\n",
       "      <td>1151</td>\n",
       "      <td>894</td>\n",
       "      <td>1171</td>\n",
       "      <td>885</td>\n",
       "      <td>1167</td>\n",
       "      <td>179.786408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ligule4</td>\n",
       "      <td>B100_rep1_d10_plm2</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>786</td>\n",
       "      <td>1423</td>\n",
       "      <td>789</td>\n",
       "      <td>1402</td>\n",
       "      <td>772</td>\n",
       "      <td>1404</td>\n",
       "      <td>51.358025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>leaf4</td>\n",
       "      <td>B100_rep1_d10_plm3</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>571</td>\n",
       "      <td>1344</td>\n",
       "      <td>592</td>\n",
       "      <td>1336</td>\n",
       "      <td>593</td>\n",
       "      <td>1342</td>\n",
       "      <td>136.758621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ligule2</td>\n",
       "      <td>B100_rep1_d10_plm4</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>793</td>\n",
       "      <td>1523</td>\n",
       "      <td>796</td>\n",
       "      <td>1512</td>\n",
       "      <td>783</td>\n",
       "      <td>1519</td>\n",
       "      <td>77.395349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>leaf2</td>\n",
       "      <td>B100_rep1_d10_plm5</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>712</td>\n",
       "      <td>1511</td>\n",
       "      <td>735</td>\n",
       "      <td>1508</td>\n",
       "      <td>734</td>\n",
       "      <td>1512</td>\n",
       "      <td>165.041667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>base</td>\n",
       "      <td>B100_rep1_d10_plm6</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>803</td>\n",
       "      <td>1555</td>\n",
       "      <td>796</td>\n",
       "      <td>1538</td>\n",
       "      <td>807</td>\n",
       "      <td>1535</td>\n",
       "      <td>205.491525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>leaf3</td>\n",
       "      <td>B100_rep1_d10_plm7</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>922</td>\n",
       "      <td>1415</td>\n",
       "      <td>898</td>\n",
       "      <td>1420</td>\n",
       "      <td>900</td>\n",
       "      <td>1416</td>\n",
       "      <td>133.183673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ligule3</td>\n",
       "      <td>B100_rep1_d10_plm8</td>\n",
       "      <td>B100_rep1_d10</td>\n",
       "      <td>800</td>\n",
       "      <td>1477</td>\n",
       "      <td>816</td>\n",
       "      <td>1459</td>\n",
       "      <td>801</td>\n",
       "      <td>1454</td>\n",
       "      <td>48.893258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>leaf5</td>\n",
       "      <td>B100_rep1_d11_plm1</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>912</td>\n",
       "      <td>1123</td>\n",
       "      <td>905</td>\n",
       "      <td>1142</td>\n",
       "      <td>896</td>\n",
       "      <td>1139</td>\n",
       "      <td>201.460784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ligule5</td>\n",
       "      <td>B100_rep1_d11_plm2</td>\n",
       "      <td>B100_rep1_d11</td>\n",
       "      <td>786</td>\n",
       "      <td>1370</td>\n",
       "      <td>792</td>\n",
       "      <td>1350</td>\n",
       "      <td>783</td>\n",
       "      <td>1347</td>\n",
       "      <td>91.428571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     group             plmname       filename  plm_x  plm_y  SS_x  SS_y  TS_x  \\\n",
       "0    leaf5  B100_rep1_d10_plm1  B100_rep1_d10    901   1151   894  1171   885   \n",
       "1  ligule4  B100_rep1_d10_plm2  B100_rep1_d10    786   1423   789  1402   772   \n",
       "2    leaf4  B100_rep1_d10_plm3  B100_rep1_d10    571   1344   592  1336   593   \n",
       "3  ligule2  B100_rep1_d10_plm4  B100_rep1_d10    793   1523   796  1512   783   \n",
       "4    leaf2  B100_rep1_d10_plm5  B100_rep1_d10    712   1511   735  1508   734   \n",
       "5     base  B100_rep1_d10_plm6  B100_rep1_d10    803   1555   796  1538   807   \n",
       "6    leaf3  B100_rep1_d10_plm7  B100_rep1_d10    922   1415   898  1420   900   \n",
       "7  ligule3  B100_rep1_d10_plm8  B100_rep1_d10    800   1477   816  1459   801   \n",
       "8    leaf5  B100_rep1_d11_plm1  B100_rep1_d11    912   1123   905  1142   896   \n",
       "9  ligule5  B100_rep1_d11_plm2  B100_rep1_d11    786   1370   792  1350   783   \n",
       "\n",
       "   TS_y    CC_ratio  \n",
       "0  1167  179.786408  \n",
       "1  1404   51.358025  \n",
       "2  1342  136.758621  \n",
       "3  1519   77.395349  \n",
       "4  1512  165.041667  \n",
       "5  1535  205.491525  \n",
       "6  1416  133.183673  \n",
       "7  1454   48.893258  \n",
       "8  1139  201.460784  \n",
       "9  1347   91.428571  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "landmark_feat_standards = pd.read_csv('B100_timeseries_test_plms_annotated.csv')\n",
    "landmark_feat_standards.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dea09cdf-2e38-4a18-bcc0-170e764e69a6",
   "metadata": {},
   "source": [
    "After glancing at the table above we essentially have 3 types of features we're classifying, our leaf tips denoted as 'leaf', our axils where leaf blades attach to the stem as 'ligule' (common term for this feature in grasses), and 'base' which represents the bottom landmark at the base of our plant.  Now we have a list of known features which we can compare to our corresponding list of our predicted homology groups.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4506c3fd-9863-44cc-a3b2-755ecd9faba2",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexingError",
     "evalue": "Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexingError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-b42b63acf88b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpcv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhomology\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstellaqc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdenovo_groups\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlandmark_pandas\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mannotated_groups\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlandmark_feat_standards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/plantcv-3.13.0+317.gc8af320a-py3.7.egg/plantcv/plantcv/homology/constellaqc.py\u001b[0m in \u001b[0;36mconstellaqc\u001b[0;34m(denovo_groups, annotated_groups)\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0manno\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mknown_feat\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0;31m# anno_bool_index = annotated_groups.loc[:, 'group'] == anno\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m         \u001b[0manno_group_calls\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdenovo_groups\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mannotated_groups\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'group'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0manno\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'group'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m         \u001b[0;31m# print(anno, 'count: ', np.sum(anno_bool_index))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m         \u001b[0mscore_row\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    887\u001b[0m                     \u001b[0;31m# AttributeError for IntervalTree get_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    888\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 889\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    890\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    891\u001b[0m             \u001b[0;31m# we by definition only have the 0th axis\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1067\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_multi_take\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1068\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1069\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_tuple_same_dim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1070\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1071\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_tuple_same_dim\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m    773\u001b[0m                 \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    774\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 775\u001b[0;31m             \u001b[0mretval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretval\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    776\u001b[0m             \u001b[0;31m# We should never have retval.ndim < self.ndim, as that should\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    777\u001b[0m             \u001b[0;31m#  be handled by the _getitem_lowerdim call above.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1102\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_slice_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1103\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_bool_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1104\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getbool_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1105\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mis_list_like_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getbool_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m    910\u001b[0m         \u001b[0;31m# caller is responsible for ensuring non-None axis\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    911\u001b[0m         \u001b[0mlabels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 912\u001b[0;31m         \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_bool_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    913\u001b[0m         \u001b[0minds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    914\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_take_with_is_copy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plantcv/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36mcheck_bool_indexer\u001b[0;34m(index, key)\u001b[0m\n\u001b[1;32m   2268\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2269\u001b[0m             raise IndexingError(\n\u001b[0;32m-> 2270\u001b[0;31m                 \u001b[0;34m\"Unalignable boolean Series provided as \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2271\u001b[0m                 \u001b[0;34m\"indexer (index of the boolean Series and of \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2272\u001b[0m                 \u001b[0;34m\"the indexed object do not match).\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexingError\u001b[0m: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match)."
     ]
    }
   ],
   "source": [
    "pcv.homology.constellaqc(denovo_groups=landmark_pandas, annotated_groups=landmark_feat_standards)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1de9e24f-8609-4282-82b4-687a302daacd",
   "metadata": {},
   "source": [
    "And there we have it! As expected the valid calls were perfect within this tutorial although error, and importantly the type of error is important to keep track of when developing this workflow for your own research.  To provide a bit of context let's discuss what our two sources of error represent.\n",
    "\n",
    "## Splitting Error\n",
    "\n",
    "Splitting errors are essentially calls in which more than one de novo homology group was generated to represent a single, known, feature.  Within this workflow these errors are often considered less egregious given that they can easily be reconciled together during manual curation of homology groups prior to using plm homology groups for morphometric analyses.  A good analogy to this problem is that of scaffold generation during whole genome sequencing in which often only fragments of rather than complete chromosomes are reconstructed from the data.  This issue is easily reconciled by a user specifying that these two scaffolds belong together and manually assigning linkage based on known attributes of this data which exist beyond the capacity of the de novo assembler.  In a similar vein of logic, if a leaf tip is broken into two groups it can easily be tied together as these groups are given a biologically relevant name.\n",
    "\n",
    "## Clumping Error\n",
    "\n",
    "Clumping errors by contrast are calls in which multiple known features are linked together under a single de novo homology group.  Understandably this error is considered far more troubling and all efforts in designing this workflow have been to drive this error rate as low as possible (in most cases hovering in the 5% range for true experimental data).  Often datasets which have a high degree of parallax (possessing perspective related artifacts of compressing 3-dimensional structures into a 2-dimensional frame) tend to drive up this error rate.  It is often best to check this error rate under a reduced dataset of each genotype or environmental treatment that is anticipated to be used given that it can provide a user with an overall grasp of how well Constella's de novo assignments work within this pool of the data.  Cases in which clumping error rates are higher may require a more stringent round of manual curration in order to ensure that morphometric analyses performed on this data afterwards are meaningful.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51dc6f7b-8e33-434e-b3e2-340b8e0d739c",
   "metadata": {},
   "source": [
    "# In Conclusion\n",
    "\n",
    "And there you have it! We've successfully started with a handful of time series images and learned how to prepare binary masks which can be used for acute in Tutorial 1.  We then learned in Tutorial 2 how to scale up what we had learned in our first exercise to work on batch image datasets. Following the generation of this batch plm data we were then able to explore de novo homology grouping through the use of our StarScape & Constella pipeline in Tutorial 3. And finally in our last exercise we again scaled up what we learned for homology grouping on to batch datasets then were able to test what we generated using ConstellaQC in order to get a general idea of how much confidence we can have in our calls.  \n",
    "\n",
    "I hope this tutorial series have been informative and provides you with some quick-start code to get your own projects running. Cheers!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a016b63-9e59-4dcd-a622-5a8210357c56",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:plantcv] *",
   "language": "python",
   "name": "conda-env-plantcv-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
